src/extension/simple_op.rs

```diff
@@ -0,0 +1,231 @@
+//! A trait that enum for op definitions that gathers up some shared functionality.
+
+use smol_str::SmolStr;
+use strum::IntoEnumIterator;
+
+use crate::{
+    ops::{custom::ExtensionOp, OpName, OpType},
+    types::TypeArg,
+    Extension,
+};
+
+use super::{
+    op_def::SignatureFunc, ExtensionBuildError, ExtensionId, ExtensionRegistry, OpDef,
+    SignatureError,
+};
+use delegate::delegate;
+use thiserror::Error;
+
+/// Error loading operation.
+#[derive(Debug, Error, PartialEq)]
+#[error("{0}")]
+#[allow(missing_docs)]
+pub enum OpLoadError {
+    #[error("Op with name {0} is not a member of this set.")]
+    NotMember(String),
+    #[error("Type args invalid: {0}.")]
+    InvalidArgs(#[from] SignatureError),
+}
+
+impl<T> OpName for T
+where
+    for<'a> &'a T: Into<&'static str>,
+{
+    fn name(&self) -> SmolStr {
+        let s = self.into();
+        s.into()
+    }
+}
+
+/// Traits implemented by types which can add themselves to [`Extension`]s as
+/// [`OpDef`]s or load themselves from an [`OpDef`]
+pub trait MakeOpDef: OpName {
+    /// Try to load one of the operations of this set from an [OpDef].
+    fn from_def(op_def: &OpDef) -> Result<Self, OpLoadError>
+    where
+        Self: Sized;
+
+    /// Return the signature (polymorphic function type) of the operation.
+    fn signature(&self) -> SignatureFunc;
+
+    /// Description of the operation. By default, the same as `self.name()`.
+    fn description(&self) -> String {
+        self.name().to_string()
+    }
+
+    /// Edit the opdef before finalising. By default does nothing.
+    fn post_opdef(&self, _def: &mut OpDef) {}
+
+    /// Add an operation implemented as an [MakeOpDef], which can provide the data
+    /// required to define an [OpDef], to an extension.
+    fn add_to_extension(&self, extension: &mut Extension) -> Result<(), ExtensionBuildError> {
+        let def = extension.add_op(self.name(), self.description(), self.signature())?;
+
+        self.post_opdef(def);
+
+        Ok(())
+    }
+
+    /// Load all variants of an enum of op definitions in to an extension as op defs.
+    /// See [strum::IntoEnumIterator].
+    fn load_all_ops(extension: &mut Extension) -> Result<(), ExtensionBuildError>
+    where
+        Self: IntoEnumIterator,
```

Comment:
`MakeOpDef` is still lightly tied to strum.

Perhaps we could still mention plain enums as an example usecase in the trait description?

Reply:
yes, do you think it's worth putting this function (and maybe others) behind a feature flag and making strum a feature dependency?

Reply:
It doesn't add many deps;
```
├── strum v0.25.0
├── strum_macros v0.25.3 (proc-macro)
│   ├── heck v0.4.1
│   ├── proc-macro2 v1.0.69 (*)
│   ├── quote v1.0.33 (*)
│   ├── rustversion v1.0.14 (proc-macro)
│   └── syn v2.0.38 (*)
```
(from `cargo-tree`, asterisks are already-required deps).

I'd leave it be for now.